File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;

class Solution {
    int t;
    int answer = 0;
    
    public void dfs(int[] numbers, int idx, boolean b, int sum){
        if(b){
            sum -= numbers[idx];
        } else {
            sum += numbers[idx];
        }
        
        if(idx == numbers.length - 1){
            if(sum == t){
                answer++;
            }
            return;
        } else {
            dfs(numbers, idx+1, !b, sum);
            dfs(numbers, idx+1, b, sum);
        }
    }
    
    public int solution(int[] numbers, int target) {
        t = target;
        
        dfs(numbers, 0, false, 0);
        dfs(numbers, 0, true, 0);
        
        return answer;
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.kt
class Solution {
    
    var answer = 0
    
    fun findTargetNumberWays(numbers: IntArray, target: Int): Int {
    return dfs(numbers, target, 0, 0)
}

    fun dfs(numbers: IntArray, target: Int, index: Int, sum: Int): Int {
        if (index == numbers.size) {
            return if (sum == target) 1 else 0
        }

        return dfs(numbers, target, index + 1, sum + numbers[index]) +
               dfs(numbers, target, index + 1, sum - numbers[index])
    }
    
    fun solution(numbers: IntArray, target: Int): Int {
        
        val result = findTargetNumberWays(numbers, target)
        
        return result
    }
}